src/stdlib/Json.ts

```diff
@@ -0,0 +1,127 @@
+import { AssociativeArray } from "../brsTypes/components/AssociativeArray";
+import { BrsArray } from "../brsTypes/components/BrsArray";
+import { Interpreter } from "../interpreter";
+import { Literal } from "../parser/Expression";
+import { randomBytes } from "crypto";
+
+import {
+    BrsBoolean,
+    BrsInvalid,
+    BrsString,
+    BrsType,
+    Callable,
+    Float,
+    Int32,
+    Int64,
+    ValueKind
+} from "../brsTypes";
+
+function brsValueOf(x: any): any {
+    if (x === null) { return BrsInvalid.Instance; }
+    let t: string = typeof x;
+    switch (t) {
+    case "boolean":
+        return BrsBoolean.from(x);
+    case "string":
+        return new BrsString(x);
+    case "number":
+        if (Number.isInteger(x)) {
+            return x >= -2_147_483_648 && x <= 2_147_483_647 ? new Int32(x) : new Int64(x);
+        }
+        return new Float(x);
+    case "object":
+        if (Array.isArray(x)) {
+            return new BrsArray(x.map(brsValueOf));
+        }
+        return new AssociativeArray(
+            Object.getOwnPropertyNames(x).map((k: string) => {
+                return {
+                    name: brsValueOf(k),
+                    value: brsValueOf(x[k])
+                };
+            })
+        );
+    default:
+        throw new Error(`brsValueOf not implemented for: ${x} <${t}>`);
+    }
+}
+
+function jsonOf(interpreter: Interpreter, x: BrsType, uid: BrsString): string {
+    switch (x.kind) {
+    case ValueKind.Invalid:
+        return "null";
+    case ValueKind.String:
+        return `"${x.toString()}"`;
+    case ValueKind.Boolean:
+    case ValueKind.Double:
+    case ValueKind.Float:
+    case ValueKind.Int32:
+    case ValueKind.Int64:
+        return x.toString();
+    case ValueKind.Object:
+        if (x instanceof AssociativeArray) {
+            try {
+                if (x.get(uid) !== BrsInvalid.Instance) {
+                    throw new Error("Nested object reference");
+                }
+                let elements = x.getElements();
+                x.set(uid, BrsBoolean.True);
```

Comment:
This is a super neat approach!  I'm unfortunately not comfortable modifying elements in a formatting function.  Regardless of my comfort, attempting to set a random property won't work on a BrightScript array, which suffers from the same "Nested object reference" problem that associative arrays do, e.g.:

```brightscript
sub main()
    a = [1, 2, 3]
    a.push(a)
    print formatJson(a)
end sub
```

This seems like a good place to use a JavaScript [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) to keep track of the objects that have been visited so far.  Something like this should work:

```typescript
/**
  * Creates a JavaScript string containing the JSON representation of the BrightScript value `x`.
  * @param x the thing
  * @returns a string (oh look that rhymed!)
  */
function jsonOf(x: BrsType): string {
    let visitedElements = new Set();

    /** Function-local implementation, used to close over `visitedElements`.  Should only be called from within `__jsonOf`, recursively. */
    function __jsonOf(x: brsType): string {
        if (visitedElements.has(x)) {
            // duplicate element detected!
            // throw an error.
        }
        visitedElements.add(x);
        switch (x.kind) {
            // most of what you have already
        }
    }

    return __jsonOf(x);
}
```

Reply:
Done, the Set approach is way cleaner. I had forgotten that JS keys can be any object not just strings. BrightScript is rotting my brain!